man/man3/TCP.3
.TH "TCP" 3 "Thu Dec 10 2020" "Version 01" "Shared library Connection Socket" \" -*- nroff -*-
.ad l
.nh
.SH NAME
TCP \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <TCP\&.h>\fP
.PP
Inherits \fBConnection\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBTCP\fP (uint32_t address, uint16_t port)"
.br
.ti -1c
.RI "\fBTCP\fP (const char *address, uint16_t port)"
.br
.ti -1c
.RI "\fBTCP\fP (std::string const &address, uint16_t port)"
.br
.ti -1c
.RI "\fB~TCP\fP () override"
.br
.ti -1c
.RI "template<typename T > std::size_t \fBReceive\fP (T *value, std::size_t tu_size)"
.br
.RI "\fIReceive value Virtual receiving function for values\&. \fP"
.ti -1c
.RI "template<typename T > std::size_t \fBSend\fP (T const *value, std::size_t tu_size)"
.br
.RI "\fISend value Virtual sending function for values\&. \fP"
.ti -1c
.RI "template<> std::size_t \fBReceive\fP (std::string *value, std::size_t const tu_size)"
.br
.ti -1c
.RI "template<> std::size_t \fBSend\fP (std::string const *value, std::size_t const tu_size)"
.br
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
Definition at line 6 of file TCP\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "TCP::TCP (uint32_t address, uint16_t port)\fC [explicit]\fP"

.PP
Definition at line 4 of file TCP\&.cpp\&.
.SS "TCP::TCP (const char * address, uint16_t port)\fC [explicit]\fP"

.PP
Definition at line 15 of file TCP\&.cpp\&.
.SS "TCP::TCP (std::string const & address, uint16_t port)\fC [explicit]\fP"

.PP
Definition at line 26 of file TCP\&.cpp\&.
.SS "TCP::~TCP ()\fC [override]\fP"

.PP
Definition at line 37 of file TCP\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "template<typename T > template std::size_t TCP::Receive (T * value, std::size_t tu_size)"

.PP
Receive value Virtual receiving function for values\&. 
.PP
\fBParameters:\fP
.RS 4
\fIvalue\fP Address of value to receive 
.br
\fItu_size\fP Bytes to receive 
.RE
.PP
\fBReturns:\fP
.RS 4
Count of received bytes 
.RE
.PP

.PP
Definition at line 46 of file TCP\&.cpp\&.
.SS "template<> std::size_t TCP::Receive (std::string * value, std::size_t const tu_size)"

.PP
Definition at line 135 of file TCP\&.cpp\&.
.SS "template<typename T > template std::size_t TCP::Send (T const * value, std::size_t tu_size)"

.PP
Send value Virtual sending function for values\&. 
.PP
\fBParameters:\fP
.RS 4
\fIvalue\fP Address of value to send 
.br
\fItu_size\fP Bytes to send 
.RE
.PP
\fBReturns:\fP
.RS 4
Count of sent bytes 
.RE
.PP
\fBTCP\fP MTU = 1460 net\&.core\&.rmem_max = 212992
.PP
Definition at line 74 of file TCP\&.cpp\&.
.SS "template<> std::size_t TCP::Send (std::string const * value, std::size_t const tu_size)"

.PP
Definition at line 141 of file TCP\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for Shared library Connection Socket from the source code\&.


src/TCP.cpp
#include "TCP.h"
#include "type_name.h"

TCP::TCP(uint32_t address, uint16_t port)
		: Connection(_TCP, address, port) {
#ifndef NDEBUG
	debug_mutex.lock();
	std::clog << "[SOCK_CONNECT] TCP::TCP("
			<< type_name<decltype(address)>() << " address: " << address << ", "
			<< type_name<decltype(port)>() << " port: " << port << ")" << '\n' << std::flush;
	debug_mutex.unlock();
#endif
}

TCP::TCP(const char *address, uint16_t port)
		: Connection(_TCP, address, port) {
#ifndef NDEBUG
	debug_mutex.lock();
	std::clog << "[SOCK_CONNECT] TCP::TCP("
			<< type_name<decltype(address)>() << " address: " << address << ", "
			<< type_name<decltype(port)>() << " port: " << port << ")" << '\n' << std::flush;
	debug_mutex.unlock();
#endif
}

TCP::TCP(const std::string &address, uint16_t port)
		: Connection(_TCP, address.data(), port) {
#ifndef NDEBUG
	debug_mutex.lock();
	std::clog << "[SOCK_CONNECT] TCP::TCP("
			<< type_name<decltype(address)>() << " address: " << address << ", "
			<< type_name<decltype(port)>() << " port: " << port << ")" << '\n' << std::flush;
	debug_mutex.unlock();
#endif
}

TCP::~TCP() {
#ifndef NDEBUG
	debug_mutex.lock();
	std::clog << "[SOCK_CONNECT] TCP::~TCP()\n" << std::flush;
	debug_mutex.unlock();
#endif
}

template <typename T>
std::size_t TCP::Receive(T *value, std::size_t const tu_size) {
	auto recv_left = tu_size;
	std::size_t total = 0;
	while (total < tu_size) {
		msg_sz = recv(get_descriptor(), value + total, recv_left, MSG_NOSIGNAL);
		if (msg_sz < 1) {
#ifndef NDEBUG
			debug_mutex.lock();
			std::clog << "[SOCK_CONNECT] TCP::Receive<" << type_name<decltype(value)>()
					  << ">(fd: " << get_descriptor() << "): DISCONNECTED" << '\n' << std::flush;
			debug_mutex.unlock();
#endif
			this->state = false;
			return total;
		}
		recv_left -= static_cast<unsigned long>(msg_sz);
		total += static_cast<unsigned long>(msg_sz);
	}
#ifndef NDEBUG
	debug_mutex.lock();
	std::clog << "[SOCK_CONNECT] TCP::Receive<" << type_name<decltype(value)>() << ">: <" << print_values(value, tu_size)
			  << " [+" << total << "]>\n" << std::flush;
	debug_mutex.unlock();
#endif
	return total;
}

template <typename T>
std::size_t TCP::Send(T const *value, std::size_t const tu_size) {
	/**
	 * TCP MTU = 1460
	 * net.core.rmem_max = 212992
	 */
	auto send_left = tu_size;
	std::size_t total = 0;
	while (send_left > 0) {
		msg_sz = send(get_descriptor(), value + total, send_left, MSG_NOSIGNAL);
		if (msg_sz < 1) {
#ifndef NDEBUG
			debug_mutex.lock();
			std::clog << "[SOCK_CONNECT] TCP::Send<" << type_name<decltype(value)>()
					  << ">(fd: " << get_descriptor() << "): DISCONNECTED" << '\n' << std::flush;
			debug_mutex.unlock();
#endif
			this->state = false;
			return total;
		}
		send_left -= static_cast<unsigned long>(msg_sz);
		total += static_cast<unsigned long>(msg_sz);
	}
#ifndef NDEBUG
	debug_mutex.lock();
	std::clog << "[SOCK_CONNECT] TCP::Send<" << type_name<decltype(value)>() << ">: <" << print_values(value, tu_size)
			  << " [+" << total << "]>\n" << std::flush;
	debug_mutex.unlock();
#endif
	return total;
}

template std::size_t TCP::Receive(char *, std::size_t);
template std::size_t TCP::Receive(unsigned char *, std::size_t);
template std::size_t TCP::Receive(short int *, std::size_t);
template std::size_t TCP::Receive(unsigned short int *, std::size_t);
template std::size_t TCP::Receive(int *, std::size_t);
template std::size_t TCP::Receive(unsigned int *, std::size_t);
template std::size_t TCP::Receive(long int *, std::size_t);
template std::size_t TCP::Receive(unsigned long int *, std::size_t);
template std::size_t TCP::Receive(long long int *, std::size_t);
template std::size_t TCP::Receive(unsigned long long int *, std::size_t);
template std::size_t TCP::Receive(float *, std::size_t);
template std::size_t TCP::Receive(double *, std::size_t);
template std::size_t TCP::Receive(long double *, std::size_t);
template std::size_t TCP::Receive(bool *, std::size_t);
template std::size_t TCP::Send(char const *, std::size_t);
template std::size_t TCP::Send(unsigned char const *, std::size_t);
template std::size_t TCP::Send(short int const *, std::size_t);
template std::size_t TCP::Send(unsigned short int const *, std::size_t);
template std::size_t TCP::Send(int const *, std::size_t);
template std::size_t TCP::Send(unsigned int const *, std::size_t);
template std::size_t TCP::Send(long int const *, std::size_t);
template std::size_t TCP::Send(unsigned long int const *, std::size_t);
template std::size_t TCP::Send(long long int const *, std::size_t);
template std::size_t TCP::Send(unsigned long long int const *, std::size_t);
template std::size_t TCP::Send(float const *, std::size_t);
template std::size_t TCP::Send(double const *, std::size_t);
template std::size_t TCP::Send(long double const *, std::size_t);
template std::size_t TCP::Send(bool const *, std::size_t);

template <>
std::size_t TCP::Receive(std::string *value, std::size_t const tu_size) {
    value->resize(tu_size < value->max_size() ? tu_size : value->max_size(), '\0');
	return TCP::Receive(&value->front(), tu_size);
}

template <>
std::size_t TCP::Send(std::string const *value, std::size_t const tu_size) {
	return TCP::Send(value->c_str(), tu_size);
}
